cpp/src/arrow/util/future.h

```diff
@@ -264,11 +335,125 @@ class Future {
+  /// callbacks. If a callback returns:
+  /// - void, a Future<> will be produced which will completes successully as soon
+  ///   as the callback runs.
+  /// - Status, a Future<> will be produced which will complete with the returned Status
+  ///   as soon as the callback runs.
+  /// - V or Result<V>, a Future<V> will be produced which will complete with the result
+  ///   of invoking the callback as soon as the callback runs.
+  /// - Future<V>, a Future<V> will be produced which will be marked complete when the
+  ///   future returned by the callback completes (and will complete with the same
+  ///   result).
+  ///
+  /// The continued Future type must be the same for both callbacks.
+  ///
+  /// Note that OnFailure can swallow errors, allowing continued Futures to successully
+  /// complete even if this Future fails.
+  ///
+  /// If this future is already completed then the callback will be run immediately
+  /// (before this method returns) and the returned future may already be marked complete
+  /// (it will definitely be marked complete if the callback returns a non-future or a
+  /// completed future).
+  template <typename OnSuccess, typename OnFailure,
+            typename ContinuedFuture =
+                detail::ContinueFuture::ForSignature<OnSuccess && (const T&)>>
+  ContinuedFuture Then(OnSuccess&& on_success, OnFailure&& on_failure) const {
+    static_assert(
+        std::is_same<detail::ContinueFuture::ForSignature<OnFailure && (const Status&)>,
+                     ContinuedFuture>::value,
+        "OnSuccess and OnFailure must continue with the same future type");
+
+    auto next = ContinuedFuture::Make();
+
+    struct Callback {
+      void operator()(const Result<T>& result) && {
+        if (ARROW_PREDICT_TRUE(result.ok())) {
+          detail::Continue(std::move(next), std::move(on_success), result.ValueOrDie());
+        } else {
+          detail::Continue(std::move(next), std::move(on_failure), result.status());
+        }
+      }
+
+      OnSuccess on_success;
+      OnFailure on_failure;
+      ContinuedFuture next;
+    };
+
+    AddCallback(Callback{std::forward<OnSuccess>(on_success),
+                         std::forward<OnFailure>(on_failure), next});
+
+    return next;
+  }
+
+  /// \brief Overload without OnFailure. Failures will be passed through unchanged.
+  template <typename OnSuccess,
+            typename ContinuedFuture =
+                detail::ContinueFuture::ForSignature<OnSuccess && (const T&)>>
+  ContinuedFuture Then(OnSuccess&& on_success) const {
+    return Then(std::forward<OnSuccess>(on_success), [](const Status& s) {
+      return Result<typename ContinuedFuture::ValueType>(s);
+    });
+  }
+
+  template <typename OnComplete,
+            typename ContinuedFuture = typename detail::ContinueFuture::ForSignature<
+                OnComplete && (const Status&)>>
+  ContinuedFuture Then(decltype(StatusOnly), OnComplete&& on_complete) const {
```

Comment:
Hmm, looking at the tests, I think I understand. The `StatusOnly` idiom doesn't look terrific. How about instead declaring this as `ContinuedFuture ThenBoth(OnComplete&& on_complete)`, where `OnComplete` would take either a `Result<T>` or a `Status`, depending on the future type?

Reply:
I agree with Antoine although I don't know if ThenBoth is much clearer.  The main concern I have is that marking something "StatusOnly" can cause the error state of the chain to be cleared.  For example, in RX, there are three types of callbacks.

on_success - continues the chain with success
on_failure - intercepts a failure in the chain and (attempts to) convert it to a success
on_completed - intercepts success or failure but does not convert to success

on_failure is equivalent to try/catch while on_completed is equivalent to try/finally.

My concern here is that StatusOnly makes me think of on_completed and so it can't clear the error state of the future.  However, looking at the tests, it appears that it does actually clear the error state.

Could we change the name to ThenFinally or WhenCompleted and then change the behavior so that a "StatusOnly" callback can't clear the error state?

Reply:
I still don't understand what "StatusOnly" is supposed to be for if we change this to a separate name such as "ThenBoth" or "ThenFinally".

(note the problem with "ThenFinally" is that it sounds like a last cleanup step (is it executed out of order?))

Reply:
After talking with Ben we are just going to drop StatusOnly until we have more compelling use cases.  In my mind it would be used for cleanup steps like you described (I'm not sure what you mean by out of order).

For example...

```
char * buffer = new char[1024*1024];
auto fut = LoadBuffer(buffer).Then([buffer] { return ProcessBuffer(buffer); }).ThenFinally([buffer] {delete buffer;});
```

So `ThenFinally` would be called after `LoadBuffer` and `ProcessBuffer` and it would be called regardless of whether they succeeded or failed and it would not affect the final result.  It would be equivalent to...

```
.Then((const Result<ProcessedBuffer>& result) [buffer] { delete buffer; return result; }, (const Status& error_status) [buffer] { delete buffer; return error_status; })
```

Reply:
Actually, just for completeness sake, the equivalent call to...
```
.ThenFinally([buffer] { delete buffer; })
```
...would be...
```
.Then(
  (const Result<ProcessedBuffer>& result) [buffer] { delete buffer; return result; },
  (const Status& error_status) [buffer] { delete buffer; return Result<ProcessedBuffer>(error_status); }
)
```
The key difference being that the on_failure callback also has to return `Result<ProcessedBuffer>` and can't just return `Status`.

Reply:
StatusOnly has been removed for now since it isn't used outside the tests. When a consumer is interested in receiving only completion statuses we can revisit this method